cve_bin_tool/helper_script.py

```diff
@@ -251,43 +284,129 @@ def output(self):
         rprint(self.vendor_product)
 
         self.CONSOLE.rule()
 
 
-def main(filenames, product_name=None):
+def main(argv=None):
+
+    argv = argv or sys.argv
+
+    # Reset logger level to warning
+    LOGGER.setLevel("WARNING")
+
+    parser = argparse.ArgumentParser(
+        prog="helper-script",
+        description=textwrap.dedent(
+            """
+                Helps contributors who want to write a new cve-bin-tool checker find common filenames,
+                version strings, and other necessary data for building a binary checker
+                """
+        ),
+    )
+    # scan directory args
+    parser.add_argument(
+        "filenames",
+        help="files to scan",
+        nargs="+",
+    )
+
+    # product-name args
+    parser.add_argument(
+        "--product",
+        help="provide product-name that would be searched",
+        dest="product_name",
+        action="store",
+    )
+
+    # version-name args
+    parser.add_argument(
+        "--version",
+        help="provide version that would be searched",
+        dest="version_number",
+        action="store",
+    )
+
+    # # search binary args
+    # parser.add_argument()
+
+    # log level args
+    parser.add_argument(
+        "-l",
+        "--log",
+        help="log level (default: warning)",
+        dest="log_level",
+        action="store",
+        choices=["debug", "info", "warning", "error", "critical"],
+    )
+
+    # contains-patterns string length args
+    parser.add_argument(
+        "--string-length",
+        help="changes the output string-length for CONTAINS_PATTERNS",
+        type=int,
+        action="store",
+    )
+
+    defaults = {
+        "filenames": [],
+        "product_name": None,
+        "version_number": None,
+        "log_level": "warning",
+        "string_length": 40,
+    }
+
+    with ErrorHandler(mode=ErrorMode.NoTrace):
+        raw_args = parser.parse_args(argv[1:])
+        args = {key: value for key, value in vars(raw_args).items() if value}
+
+    args = ChainMap(args, defaults)
+
+    if args["log_level"]:
+        LOGGER.setLevel(args["log_level"].upper())
+
+    filenames = args["filenames"]
+    product_name = args["product_name"]
+    version_number = args["version_number"]
+    string_length = args["string_length"]
+
     # finds common strings across multiple files for CONTAIN_PATTERNS
-    hs = HelperScript(filenames[1], product_name=product_name)
-    binary_string_list_1 = hs.extract_and_parse_file(filenames[1])
+    hs = HelperScript(
+        filenames[0],
+        product_name=product_name,
+        version_number=version_number,
+        string_length=string_length,
+    )
+    binary_string_list_1 = hs.extract_and_parse_file(filenames[0])
     binary_string_list_2 = []
 
-    for filename in filenames[2:]:
+    for filename in filenames[1:]:
         hs = HelperScript(filename)
         binary_string_list_2 = hs.extract_and_parse_file(filename)
 
```

Comment:
This whole block (lines 400-419) looks pretty complicated (variables with `_1`, `_2` in names are a clear sign of it).

On line 400 you create a `HelperScript` instance `hs` with all arguments set, however in the loop you overwrite this variable with `HelperScript(filename)`. Is it okay that you don't pass `product_name, version_number, string_length` here? Do you even need to create new instance?

Maybe this is what you wanted?
```py
    hs = HelperScript(
        filenames[0],  # I really think this is not needed in constructor
        product_name=product_name,
        version_number=version_number,
        string_length=string_length,
    )
    # maybe change this to `contains_patterns` to mirror correct Checker attribute name
    hs.contain_patterns = hs.extract_and_parse_file(filenames[0])

    for filename in filenames[1:]:
        binary_string_list = hs.extract_and_parse_file(filename)

        if binary_string_list:
            hs.contain_patterns = list(
                set(hs.contain_patterns).intersection(set(binary_string_list))
            )
```

Reply:
Now looking back at it, this is not a clean piece of code. But in short, what it intended to do was to find the `intersection` across `CONTAINS_PATTERNS` when multiple filenames are given. 

For now I'm removing it, because of some problems, i.e., 
1. this does not check if the multiple filenames provided are for the same product
  example: `$ python3 -m cve_bin_tool.helper_script ./busybox-1.33.1.rpm ./libsndfile-1.3.2.rpm` which breaks it.
2. This creates multiple instances of `HelperScript`, which slows the tool significantly.

But, this would be a nice feature to have, so, I would add this as a future goal for helper-script in an issue :)

Reply:
Can you create an issue for that if it's still something planned?

Reply:
Sorry, I almost forgot about it. I'll do that now :)